tests/deploy/test_scheduler.py

```diff
@@ -0,0 +1,311 @@
+
+    result = await client.create_project("env-test")
+    project_id = result.result["project"]["id"]
+
+    result = await client.create_environment(project_id=project_id, name="dev")
+    env_id = result.result["environment"]["id"]
+    env = await data.Environment.get_by_id(uuid.UUID(env_id))
+    await env.set(data.AUTO_DEPLOY, False)
+    await env.set(data.PUSH_ON_AUTO_DEPLOY, False)
+    await env.set(data.AGENT_TRIGGER_METHOD_ON_AUTO_DEPLOY, const.AgentTriggerMethod.push_full_deploy)
+
+    clienthelper = ClientHelper(client, env_id)
+
+    version = await clienthelper.get_version()
+
+    resources = [
+        {
+            "key": "key1",
+            "value": "value1",
+            "id": "test::Resource[agent2,key=key1],v=%d" % version,
+            "send_event": False,
+            "purged": False,
+            "requires": [],
+        },
+        {
+            "key": "key2",
+            "value": inmanta.execute.util.Unknown(source=None),
+            "id": "test::Resource[agent2,key=key2],v=%d" % version,
+            "send_event": False,
+            "purged": False,
+            "requires": [],
+        },
+        {
+            "key": "key4",
+            "value": inmanta.execute.util.Unknown(source=None),
+            "id": "test::Resource[agent2,key=key4],v=%d" % version,
+            "send_event": False,
+            "requires": ["test::Resource[agent2,key=key1],v=%d" % version, "test::Resource[agent2,key=key2],v=%d" % version],
+            "purged": False,
+        },
+        {
+            "key": "key5",
+            "value": "val",
+            "id": "test::Resource[agent2,key=key5],v=%d" % version,
+            "send_event": False,
+            "requires": ["test::Resource[agent2,key=key4],v=%d" % version],
+            "purged": False,
+        },
+    ]
+
+    status = {
+        "test::Resource[agent2,key=key4]": const.ResourceState.undefined,
+        "test::Resource[agent2,key=key2]": const.ResourceState.undefined,
+    }
+    result = await client.put_version(
+        tid=env_id,
+        version=version,
+        resources=resources,
+        resource_state=status,
+        unknowns=[],
+        version_info={},
+        compiler_version=get_compiler_version(),
+    )
+    assert result.code == 200
+
+    scheduler = ResourceScheduler()
+    await scheduler.start()
+    for resource in resources:
+        id_without_version, _, _ = resource["id"].partition(",v=")
+        assert id_without_version in scheduler._state.resources
+        expected_resource_attributes = copy.deepcopy(resource)
+        expected_resource_attributes.pop("id")
+        current_attributes = scheduler._state.resources[id_without_version].attributes
+        # TODO h ResourceScheduler -> Unknown == undefined?
+        # TODO h ResourceScheduler -> loose version information on requires?
+
+        # Already a todo for that
+        # TODO h ResourceScheduler -> undefined status lost in the process -> has update / new only?
```

Comment:
I presume that this is the desired outcome, given that the resource scheduler is intended to always work with the latest version?